src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.SslCtx.cs

```diff
@@ -69,7 +83,107 @@ protected override bool ReleaseHandle()
         {
             Interop.Ssl.SslCtxDestroy(handle);
             SetHandle(IntPtr.Zero);
+            if (_gch.IsAllocated)
+            {
+                //Interop.Ssl.SslCtxSetData(this, (IntPtr)_gch);
+                _gch.Free();
+            }
+
+            if (_sslSessions != null)
+            {
+                lock (_sslSessions)
+                {
+                    foreach (string name in _sslSessions.Keys)
+                    {
+                        _sslSessions.Remove(name, out IntPtr session);
+                        Interop.Ssl.SessionFree(session);
+                    }
+                }
+            }
+
             return true;
         }
+
+        public void EnableSessionCache()
+        {
+            _sslSessions = new ConcurrentDictionary<string, IntPtr>();
+            _gch = GCHandle.Alloc(this);
+            // This is needed so we can find the handle from session remove callback.
+            Interop.Ssl.SslCtxSetData(this, (IntPtr)_gch);
+        }
+
+        public bool TryAddSession(IntPtr serverName, IntPtr session)
+        {
+            Debug.Assert(_sslSessions != null && session != IntPtr.Zero);
+
+            if (_sslSessions == null || serverName == IntPtr.Zero)
+            {
+                return false;
+            }
+
+            string? name = Marshal.PtrToStringAnsi(serverName);
+            if (!string.IsNullOrEmpty(name))
+            {
+                Interop.Ssl.SessionSetHostname(session, serverName);
+
+                lock (_sslSessions)
```

Comment:
The locking around _sslSessions makes sense, since you're manipulating state depending on how the dictionary performed.

But, since you're already locking it, it feels like you want a non-Concurrent dictionary.

Reply:
agreed. I was also thinking about grabbing extra reference on the session. That would allow me to use ConcurrentDictionary without locking as the session would never be released in the middle. 
Do you have preference/recommendation @bartonjs ?

Reply:
> grabbing extra reference on the session

Something like

```
Interop.Ssl.SslSessionUpRef(session);

if (!_sslSessions.TryAdd(...))
{
    // Undo the upref since it's not in the dictionary
    Interop.Ssl.SslSessionFree(session);
}
```
?  (Upref inside has a race condition with the cleanup in ReleaseHandle)

That would get a little weird since in the cleanup you'd need to call free twice, I think?

The fact that we wrote ConcurrentDictionary suggests that it gives better perf (on average) than manual locking... but if the code to interact with it is doing memory/lifetime management and it becomes unreadable with the gymnastics... then locking is better for maintainability. (If it's clean code and more performant, than by all means use upref+ConcurrentDictionary)

Reply:
yes. and then call free twice. I'm inclined to good with the lock and better maintainability as the perf does not depend on this. This happens one in while - not even for each SSL session. I started with ConcurrentDictionary but you are right - we don't need it at this point.